### PR TITLE
Sb pja listshare

### LINKED ISSUE #7: 5. As a user, I want to join an existing shopping list so I can share a shopping list with someone
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

* User enters an existing token
* Save the token to `localStorage`
* Show the user the existing list

AC:

* When the user does not already have a token in `localStorage`, on the onboarding/home screen, the user can enter an existing token
* Entering the token and hitting submit saves the token to `localStorage`, effectively giving them joint control of the list
* On submit, show an error if the token does not exist
* If they get an error message, allow them to try again or create a new list

### PR DESCRIPTION
**Pull request by erostribe**
## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

Screen Shot 2020-04-14 at 1 33 01 PM

### After

Screen Shot 2020-04-21 at 5 14 48 PM

Screen Shot 2020-04-21 at 5 15 55 PM

Screen Shot 2020-04-22 at 1 52 28 PM

Screen Shot 2020-04-22 at 1 58 39 PM

## Related Issue

This PR closes #7 

### Objective 
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

- User enters an existing token
- Save the token to localStorage
- Show the user the existing list

### Steps to accomplish to build the feature
1. User enters an existing token: 

- [x]  create form field to enter token 
- [x]   save user input to state 
- [x]   verify that state has been set to user input 

2. User submits the token: 
- [x] query the db for items in the `items` collection with the given token (based on current data structure)

3. IF the token is valid (the query results length is not equal to 0):

- [x]  Save the token to localStorage 
- [x]  Show the user the existing list

4. IF the token is NOT valid (the query results length is equal to 0): 

- [x]   show an error if the token does not exist 
- [x]   allow them to try again or 
- [x]   allow them to create a new list 

## Acceptance Criteria

AC:

1. When the user does not already have a token in localStorage, on the onboarding/home screen, the user can enter an existing token
2. Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
3. On submit, show an error if the token does not exist
4. If they get an error message, allow them to try again or create a new list

## Testing Steps / QA Criteria
1)  Git pull the branch using `git pull origin sb-pja-listshare` and `git checkout sb-pja-listshare` to switch to the branch.
2) Run `npm install` to add dependencies, then `npm start` to run the app.

TO TEST: 
1. You might need to clear your localStorage of `user_token` -> run `localStorage.removeItem('user_token')` in your browser console. Refresh the page.
2. `wile cain woo` is a valid user token.

- Given: A `user_token` exists in `localStorage` the app will show the shopping list component.
- Given: A `user_token` does not exist in `localStorage`, the app will present the user with the choice to create a token **or** search for and use a `user_token` given to them by another user. 
- Given: An invalid shared token, a warning message and the options of submitting a shared token **or** creating a new token are still available.
- Given: A valid shared token should present the user with the expected shopping list.

### CODE REVIEW

File: src/components/ShareList.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React, { useState } from 'react';
+import { db } from '../lib/firebase';
+import { USER_TOKEN } from '../constants';
+import { useSharedToken } from '../lib/useSharedToken';
+
+const ShareList = () => {
+  const [isError, toggleError] = useState(false);
+  const [tokenToCheck, setUserToken] = useState();
+  const [sharedToken, saveSharedToken] = useSharedToken(tokenToCheck);
```
Comment (skillitzimberg) [inline src/components/ShareList.js]: 'sharedToken' is not being used . . . ?

File: src/components/ShareList.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const handleSubmit = event => {
+    event.preventDefault();
+    verifySharedToken();
+    console.log(event.target.value);
```
Comment (skillitzimberg) [inline src/components/ShareList.js]: Remove `console.log`

File: src/components/navbutton.js [hunk 3]
```diff
@@ -9,3 +9,4 @@ const NavButton = props => (
 );
 
 export default NavButton;
+//thingy
```
Comment (skillitzimberg) [inline src/components/navbutton.js]: : ) `//thingy`?

File: src/constants.js [hunk 4]
```diff
@@ -1 +1,2 @@
 export const USER_TOKEN = 'user_token';
+export const ITEMS = 'items';
```
Comment (jaysonjphillips) [inline src/constants.js:2]: I'm a fan of constants and I do like your approach to adding them as needed.

**Review by SaraSweetie — APPROVED** [review]
This works as intended. I am able to add to an existing shopping list if I have the user token. I also get an error message if I input an invalid token.

**Review by espain16 — APPROVED** [review]
Your code is very easy to walk through and understand. Good job !

**Comment by jaysonjphillips** [comment]
Acceptance Criteria review:
- works as expected
- pretty clean and quick validation of the token when trying to share an existing list

Nice work!


2 (non-AC) comments on functionality, just for my own understanding (non-blocking):

1. When creating a new shopping list, an empty item is created by default
2. Clicking on a shopping list item still deletes the item. Is that still the desired behavior (I'm assuming it is since we haven't yet dealt with the rest of the CRUD operations for a shopping list item yet, but wanted to double check.

**Comment by skillitzimberg** [comment]
@jaysonjphillips, when a user first clicks `Create New List` an item is added to the db. This item only has the field `user_token`.

It would probably be better to create a `users` collection and add new tokens there, just to keep the `items` collection clean and focused on its purpose.

My thought on leaving the `delete` in place is that as we add items to the list for testing we can easily remove them rather than having to click-click-click through the Firebase console.

> Acceptance Criteria review:
> 
> * works as expected
> * pretty clean and quick validation of the token when trying to share an existing list
> 
> Nice work!
> 
> 2 (non-AC) comments on functionality, just for my own understanding (non-blocking):
> 
> 1. When creating a new shopping list, an empty item is created by default
> 2. Clicking on a shopping list item still deletes the item. Is that still the desired behavior (I'm assuming it is since we haven't yet dealt with the rest of the CRUD operations for a shopping list item yet, but wanted to double check.

**Pull request merged**